File: js/ai/src/tool.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  action,
  ActionFnArg,
  assertUnstable,
  isAction,
  stripUndefinedProps,
  z,
  type Action,
  type ActionContext,
  type ActionRunOptions,
  type JSONSchema7,
} from '@genkit-ai/core';
import type { Registry } from '@genkit-ai/core/registry';
import { parseSchema, toJsonSchema } from '@genkit-ai/core/schema';
import { setCustomMetadataAttributes } from '@genkit-ai/core/tracing';
import {
  type Part,
  type ToolDefinition,
  type ToolRequestPart,
  type ToolResponsePart,
} from './model.js';
import { MultipartToolResponseSchema } from './parts.js';
import { isExecutablePrompt, type ExecutablePrompt } from './prompt.js';

export interface Resumable<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
> {
  /**
   * respond constructs a tool response corresponding to the provided interrupt tool request
   * using the provided reply data, validating it against the output schema of the tool if
   * it exists.
   *
   * @beta
   */
  respond(
    /** The interrupt tool request to which you want to respond. */
    interrupt: ToolRequestPart,
    /**
     * The data with which you want to respond. Must conform to a tool's output schema or an
     * interrupt's input schema.
     **/
    outputData: z.infer<O>,
    options?: { metadata?: Record<string, any> }
  ): ToolResponsePart;

  /**
   * restart constructs a tool request corresponding to the provided interrupt tool request
   * that will then re-trigger the tool after e.g. a user confirms. The `resumedMetadata`
   * supplied to this method will be passed to the tool to allow for custom handling of
   * restart logic.
   *
   * @param interrupt The interrupt tool request you want to restart.
   * @param resumedMetadata The metadata you want to provide to the tool to aide in reprocessing. Defaults to `true` if none is supplied.
   * @param options Additional options for restarting the tool.
   *
   * @beta
   */
  restart(
    interrupt: ToolRequestPart,
    resumedMetadata?: any,
    options?: {
      /**
       * Replace the existing input arguments to the tool with different ones, for example
       * if the user revised an action before confirming. When input is replaced, the existing
       * tool request will be amended in the message history.
       **/
      replaceInput?: z.infer<I>;
    }
  ): ToolRequestPart;
}

/**
 * An action with a `tool` type.
 */
export type ToolAction<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
> = Action<I, O, z.ZodTypeAny, ToolRunOptions> &
  Resumable<I, O> & {
    __action: {
      metadata: {
        type: 'tool';
      };
    };
  };

/**
 * An action with a `tool.v2` type.
 */
export type MultipartToolAction<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
> = Action<
  I,
  typeof MultipartToolResponseSchema,
  z.ZodTypeAny,
  ToolRunOptions
> &
  Resumable<I, O> & {
    __action: {
      metadata: {
        type: 'tool.v2';
      };
    };
  };

/**
 * A dynamic action with a `tool` type. Dynamic tools are detached actions -- not associated with any registry.
 */
export type DynamicToolAction<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
> = Action<I, O, z.ZodTypeAny, ToolRunOptions> & {
  /** @deprecated no-op, for backwards compatibility only. */
  attach(registry: Registry): ToolAction<I, O>;
} & Resumable<I, O> & {
    __action: {
      metadata: {
        type: 'tool';
      };
    };
  };

export interface ToolRunOptions extends ActionRunOptions<z.ZodTypeAny> {
  /**
   * If resumed is supplied to a tool at runtime, that means that it was previously interrupted and this is a second
   * @beta
   **/
  resumed?: boolean | Record<string, any>;
  /** The metadata from the tool request that triggered this run. */
  metadata?: Record<string, any>;
}

/**
 * Configuration for a tool.
 */
export interface ToolConfig<I extends z.ZodTypeAny, O extends z.ZodTypeAny> {
  /** Unique name of the tool to use as a key in the registry. */
  name: string;
  /** Description of the tool. This is passed to the model to help understand what the tool is used for. */
  description: string;
  /** Input Zod schema. Mutually exclusive with `inputJsonSchema`. */
  inputSchema?: I;
  /** Input JSON schema. Mutually exclusive with `inputSchema`. */
  inputJsonSchema?: JSONSchema7;
  /** Output Zod schema. Mutually exclusive with `outputJsonSchema`. */
  outputSchema?: O;
  /** Output JSON schema. Mutually exclusive with `outputSchema`. */
  outputJsonSchema?: JSONSchema7;
  /** Metadata to be passed to the tool. */
  metadata?: Record<string, any>;
}

/**
 * A reference to a tool in the form of a name, definition, or the action itself.
 */
export type ToolArgument<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  O extends z.ZodTypeAny = z.ZodTypeAny,
> = string | ToolAction<I, O> | Action<I, O> | ExecutablePrompt<any, any, any>;

/**
 * Converts an action to a tool action by setting the appropriate metadata.
 */
export function asTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  registry: Registry,
  action: Action<I, O>
): ToolAction<I, O> {
  if (action.__action?.metadata?.type === 'tool') {
    return action as ToolAction<I, O>;
  }

  const fn = ((input) => {
    setCustomMetadataAttributes({ subtype: 'tool' });
    return action(input);
  }) as ToolAction<I, O>;
  fn.__action = {
    ...action.__action,
    metadata: { ...action.__action.metadata, type: 'tool' },
  };
  return fn;
}

/**
 * Resolves a mix of various formats of tool references to a list of tool actions by looking them up in the registry.
 */
export async function resolveTools<
  O extends z.ZodTypeAny = z.ZodTypeAny,
  CustomOptions extends z.ZodTypeAny = z.ZodTypeAny,
>(
  registry: Registry,
  tools?: (ToolArgument | ToolDefinition)[]
): Promise<ToolAction[]> {
  if (!tools || tools.length === 0) {
    return [];
  }

  return await Promise.all(
    tools.map(async (ref): Promise<ToolAction> => {
      if (typeof ref === 'string') {
        return await lookupToolByName(registry, ref);
      } else if (isAction(ref)) {
        return asTool(registry, ref);
      } else if (isExecutablePrompt(ref)) {
        return await ref.asTool();
      } else if ((ref as ToolDefinition).name) {
        return await lookupToolByName(
          registry,
          (ref as ToolDefinition).metadata?.originalName ||
            (ref as ToolDefinition).name
        );
      }
      throw new Error('Tools must be strings, tool definitions, or actions.');
    })
  );
}

export async function lookupToolByName(
  registry: Registry,
  name: string
): Promise<ToolAction> {
  const tool =
    (await registry.lookupAction(name)) ||
    (await registry.lookupAction(`/tool/${name}`)) ||
    (await registry.lookupAction(`/tool.v2/${name}`)) ||
    (await registry.lookupAction(`/prompt/${name}`)) ||
    (await registry.lookupAction(`/dynamic-action-provider/${name}`));
  if (!tool) {
    throw new Error(`Tool ${name} not found`);
  }
  return tool as ToolAction;
}

/**
 * Converts a tool action to a definition of the tool to be passed to a model.
 */
export function toToolDefinition(
  tool: Action<z.ZodTypeAny, z.ZodTypeAny>
): ToolDefinition {
  const originalName = tool.__action.name;
  let name = originalName;
  if (originalName.includes('/')) {
    name = originalName.substring(originalName.lastIndexOf('/') + 1);
  }

  const out: ToolDefinition = {
    name,
    description: tool.__action.description || '',
    outputSchema: toJsonSchema({
      schema: tool.__action.outputSchema ?? z.void(),
      jsonSchema: tool.__action.outputJsonSchema,
    })!,
    inputSchema: toJsonSchema({
      schema: tool.__action.inputSchema ?? z.void(),
      jsonSchema: tool.__action.inputJsonSchema,
    })!,
  };

  if (originalName !== name) {
    out.metadata = { originalName };
  }

  return out;
}

export interface ToolFnOptions extends ActionFnArg<never> {
  /**
   * A function that can be called during tool execution that will result in the tool
   * getting interrupted (immediately) and tool request returned to the upstream caller.
   */
  interrupt: (metadata?: Record<string, any>) => never;

  context: ActionContext;
}

export type ToolFn<I extends z.ZodTypeAny, O extends z.ZodTypeAny> = (
  input: z.infer<I>,
  ctx: ToolFnOptions & ToolRunOptions
) => Promise<z.infer<O>>;

export type MultipartToolFn<I extends z.ZodTypeAny, O extends z.ZodTypeAny> = (
  input: z.infer<I>,
  ctx: ToolFnOptions & ToolRunOptions
) => Promise<{
  output?: z.infer<O>;
  content?: Part[];
}>;

export function defineTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  registry: Registry,
  config: { multipart: true } & ToolConfig<I, O>,
  fn?: ToolFn<I, O>
): MultipartToolAction<I, O>;
export function defineTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  registry: Registry,
  config: ToolConfig<I, O>,
  fn?: ToolFn<I, O>
): ToolAction<I, O>;

/**
 * Defines a tool.
 *
 * A tool is an action that can be passed to a model to be called automatically if it so chooses.
 */
export function defineTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  registry: Registry,
  config: { multipart?: true } & ToolConfig<I, O>,
  fn?: ToolFn<I, O> | MultipartToolFn<I, O>
): ToolAction<I, O> | MultipartToolAction<I, O> {
  const a = tool(config, fn);
  delete a.__action.metadata.dynamic;
  registry.registerAction(config.multipart ? 'tool.v2' : 'tool', a);
  if (!config.multipart) {
    // For non-multipart tools, we register a v2 tool action as well
    registry.registerAction('tool.v2', basicToolV2(config, fn as ToolFn<I, O>));
  }
  return a as ToolAction<I, O>;
}

function implementTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  a: ToolAction<I, O>,
  config: ToolConfig<I, O>,
  registry?: Registry
) {
  (a as ToolAction<I, O>).respond = (interrupt, responseData, options) => {
    if (registry) {
      assertUnstable(
        registry,
        'beta',
        "The 'tool.reply' method is part of the 'interrupts' beta feature."
      );
    }
    parseSchema(responseData, {
      jsonSchema: config.outputJsonSchema,
      schema: config.outputSchema,
    });
    return {
      toolResponse: stripUndefinedProps({
        name: interrupt.toolRequest.name,
        ref: interrupt.toolRequest.ref,
        output: responseData,
      }),
      metadata: {
        interruptResponse: options?.metadata || true,
      },
    };
  };

  (a as ToolAction<I, O>).restart = (interrupt, resumedMetadata, options) => {
    if (registry) {
      assertUnstable(
        registry,
        'beta',
        "The 'tool.restart' method is part of the 'interrupts' beta feature."
      );
    }
    let replaceInput = options?.replaceInput;
    if (replaceInput) {
      replaceInput = parseSchema(replaceInput, {
        schema: config.inputSchema,
        jsonSchema: config.inputJsonSchema,
      });
    }
    return {
      toolRequest: stripUndefinedProps({
        name: interrupt.toolRequest.name,
        ref: interrupt.toolRequest.ref,
        input: replaceInput || interrupt.toolRequest.input,
      }),
      metadata: stripUndefinedProps({
        ...interrupt.metadata,
        resumed: resumedMetadata || true,
        // annotate the original input if replacing it
        replacedInput: replaceInput ? interrupt.toolRequest.input : undefined,
      }),
    };
  };
}

/** InterruptConfig defines the options for configuring an interrupt. */
export type InterruptConfig<
  I extends z.ZodTypeAny = z.ZodTypeAny,
  R extends z.ZodTypeAny = z.ZodTypeAny,
> = ToolConfig<I, R> & {
  /** requestMetadata adds additional `interrupt` metadata to the `toolRequest` generated by the interrupt */
  requestMetadata?:
    | Record<string, any>
    | ((
        input: z.infer<I>
      ) => Record<string, any> | Promise<Record<string, any>>);
};

export function isToolRequest(part: Part): part is ToolRequestPart {
  return !!part.toolRequest;
}

export function isToolResponse(part: Part): part is ToolResponsePart {
  return !!part.toolResponse;
}

export function isDynamicTool(t: unknown): t is ToolAction {
  return isAction(t) && t.__action?.metadata?.dynamic === true;
}

export function interrupt<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: InterruptConfig<I, O>
): ToolAction<I, O> {
  const { requestMetadata, ...toolConfig } = config;

  return tool<I, O>(toolConfig, async (input, { interrupt }) => {
    if (!config.requestMetadata) interrupt();
    else if (typeof config.requestMetadata === 'object')
      interrupt(config.requestMetadata);
    else interrupt(await Promise.resolve(config.requestMetadata(input)));
  });
}

export function defineInterrupt<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  registry: Registry,
  config: InterruptConfig<I, O>
): ToolAction<I, O> {
  const i = interrupt(config);
  registry.registerAction('tool', i);
  return i;
}

/**
 * Thrown when tools execution is interrupted. It's meant to be caugh by the framework, not public API.
 */
export class ToolInterruptError extends Error {
  constructor(readonly metadata?: Record<string, any>) {
    super();
    this.name = 'ToolInterruptError';
  }
}

/**
 * Interrupts current tool execution causing tool request to be returned in the generation response.
 * Should only be called within a tool.
 */
function interruptTool(registry?: Registry) {
  return (metadata?: Record<string, any>): never => {
    if (registry) {
      assertUnstable(registry, 'beta', 'Tool interrupts are a beta feature.');
    }
    throw new ToolInterruptError(metadata);
  };
}

export function tool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: { multipart: true } & ToolConfig<I, O>,
  fn?: ToolFn<I, O>
): MultipartToolAction<I, O>;
export function tool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: ToolConfig<I, O>,
  fn?: ToolFn<I, O>
): ToolAction<I, O>;

/**
 * Defines a dynamic tool. Dynamic tools are just like regular tools but will not be registered in the
 * Genkit registry and can be defined dynamically at runtime.
 */
export function tool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: { multipart?: true } & ToolConfig<I, O>,
  fn?: ToolFn<I, O> | MultipartToolFn<I, O>
): ToolAction<I, O> | MultipartToolAction<I, O> {
  return config.multipart ? multipartTool(config, fn) : basicTool(config, fn);
}

function basicTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: ToolConfig<I, O>,
  fn?: ToolFn<I, O>
): ToolAction<I, O> {
  const a = action(
    {
      ...config,
      actionType: 'tool',
      metadata: { ...(config.metadata || {}), type: 'tool', dynamic: true },
    },
    (i, runOptions) => {
      const interrupt = interruptTool(runOptions.registry);
      if (fn) {
        return fn(i, {
          ...runOptions,
          context: { ...runOptions.context },
          interrupt,
        });
      }
      return interrupt();
    }
  ) as ToolAction<I, O>;
  implementTool(a, config);
  return a;
}

function basicToolV2<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: ToolConfig<I, O>,
  fn?: ToolFn<I, O>
): MultipartToolAction<I, O> {
  return multipartTool(config, async (input, ctx) => {
    if (!fn) {
      const interrupt = interruptTool(ctx.registry);
      return interrupt();
    }
    return {
      output: await fn(input, ctx),
    };
  });
}

function multipartTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: ToolConfig<I, O>,
  fn?: MultipartToolFn<I, O>
): MultipartToolAction<I, O> {
  const a = action(
    {
      ...config,
      outputSchema: MultipartToolResponseSchema,
      actionType: 'tool.v2',
      metadata: {
        ...(config.metadata || {}),
        type: 'tool.v2',
        tool: { multipart: true },
      },
    },
    (i, runOptions) => {
      const interrupt = interruptTool(runOptions.registry);
      if (fn) {
        return fn(i, {
          ...runOptions,
          context: { ...runOptions.context },
          interrupt,
        });
      }
      return interrupt() as any; // we cast to any because `interrupt` throws.
    }
  ) as MultipartToolAction<I, O>;
  implementTool(a as any, config);
  return a;
}

/**
 * Defines a dynamic tool. Dynamic tools are just like regular tools but will not be registered in the
 * Genkit registry and can be defined dynamically at runtime.
 *
 * @deprecated renamed to {@link tool}.
 */
export function dynamicTool<I extends z.ZodTypeAny, O extends z.ZodTypeAny>(
  config: ToolConfig<I, O>,
  fn?: ToolFn<I, O>
): DynamicToolAction<I, O> {
  const t = basicTool(config, fn) as DynamicToolAction<I, O>;
  t.attach = (_: Registry) => t;
  return t;
}

File: java/ai/src/main/java/com/google/genkit/ai/Tool.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.genkit.ai;

import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Consumer;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.genkit.core.Action;
import com.google.genkit.core.ActionContext;
import com.google.genkit.core.ActionDesc;
import com.google.genkit.core.ActionRunResult;
import com.google.genkit.core.ActionType;
import com.google.genkit.core.GenkitException;
import com.google.genkit.core.JsonUtils;
import com.google.genkit.core.Registry;
import com.google.genkit.core.tracing.SpanMetadata;
import com.google.genkit.core.tracing.Tracer;

/**
 * Tool represents a function that can be called by an AI model.
 *
 * Tools allow models to interact with external systems and perform actions
 * during generation.
 *
 * @param <I>
 *            the input type
 * @param <O>
 *            the output type
 */
public class Tool<I, O> implements Action<I, O, Void> {

  private final String name;
  private final String description;
  private final Map<String, Object> inputSchema;
  private final Map<String, Object> outputSchema;
  private final BiFunction<ActionContext, I, O> handler;
  private final Map<String, Object> metadata;
  private final Class<I> inputClass;

  /**
   * Creates a new Tool.
   *
   * @param name
   *            the tool name
   * @param description
   *            the tool description
   * @param inputSchema
   *            the input JSON schema
   * @param outputSchema
   *            the output JSON schema
   * @param inputClass
   *            the input class for JSON deserialization
   * @param handler
   *            the tool handler function
   */
  public Tool(String name, String description, Map<String, Object> inputSchema, Map<String, Object> outputSchema,
      Class<I> inputClass, BiFunction<ActionContext, I, O> handler) {
    this.name = name;
    this.description = description;
    this.inputSchema = inputSchema;
    this.outputSchema = outputSchema;
    this.inputClass = inputClass;
    this.handler = handler;
    this.metadata = new HashMap<>();
    this.metadata.put("description", description);
  }

  /**
   * Creates a builder for Tool.
   *
   * @param <I>
   *            the input type
   * @param <O>
   *            the output type
   * @return a new builder
   */
  public static <I, O> Builder<I, O> builder() {
    return new Builder<>();
  }

  @Override
  public String getName() {
    return name;
  }

  @Override
  public ActionType getType() {
    return ActionType.TOOL;
  }

  @Override
  public ActionDesc getDesc() {
    return ActionDesc.builder().type(ActionType.TOOL).name(name).description(description).inputSchema(inputSchema)
        .outputSchema(outputSchema).build();
  }

  @Override
  public O run(ActionContext ctx, I input) throws GenkitException {
    SpanMetadata spanMetadata = SpanMetadata.builder().name(name).type(ActionType.TOOL.getValue()).subtype("tool")
        .build();

    String flowName = ctx.getFlowName();
    if (flowName != null) {
      spanMetadata.getAttributes().put("genkit:metadata:flow:name", flowName);
    }

    return Tracer.runInNewSpan(ctx, spanMetadata, input, (spanCtx, in) -> {
      try {
        O result = handler.apply(ctx.withSpanContext(spanCtx), in);
        return result;
      } catch (AgentHandoffException e) {
        // Re-throw agent handoff exceptions for multi-agent pattern
        throw e;
      } catch (ToolInterruptException e) {
        // Re-throw interrupt exceptions for human-in-the-loop pattern
        throw e;
      } catch (Exception e) {
        if (e instanceof GenkitException) {
          throw (GenkitException) e;
        }
        throw new GenkitException("Tool execution failed: " + e.getMessage(), e);
      }
    });
  }

  @Override
  public O run(ActionContext ctx, I input, Consumer<Void> streamCallback) throws GenkitException {
    return run(ctx, input);
  }

  @Override
  @SuppressWarnings("unchecked")
  public JsonNode runJson(ActionContext ctx, JsonNode input, Consumer<JsonNode> streamCallback)
      throws GenkitException {
    I typedInput = inputClass != null ? JsonUtils.fromJsonNode(input, inputClass) : (I) input;
    O output = run(ctx, typedInput);
    return JsonUtils.toJsonNode(output);
  }

  @Override
  public ActionRunResult<JsonNode> runJsonWithTelemetry(ActionContext ctx, JsonNode input,
      Consumer<JsonNode> streamCallback) throws GenkitException {
    JsonNode result = runJson(ctx, input, streamCallback);
    return new ActionRunResult<>(result, null, null);
  }

  @Override
  public Map<String, Object> getInputSchema() {
    return inputSchema;
  }

  @Override
  public Map<String, Object> getOutputSchema() {
    return outputSchema;
  }

  @Override
  public Map<String, Object> getMetadata() {
    return metadata;
  }

  @Override
  public void register(Registry registry) {
    registry.registerAction(ActionType.TOOL.keyFromName(name), this);
  }

  /**
   * Gets the tool description.
   *
   * @return the description
   */
  public String getDescription() {
    return description;
  }

  /**
   * Gets the input class for JSON deserialization.
   *
   * @return the input class, or null if not specified
   */
  public Class<I> getInputClass() {
    return inputClass;
  }

  /**
   * Gets the tool definition for use in model requests.
   *
   * @return the tool definition
   */
  public ToolDefinition getDefinition() {
    return new ToolDefinition(name, description, inputSchema, outputSchema);
  }

  /**
   * Constructs a tool response for an interrupted tool request.
   *
   * <p>
   * This method is used when resuming generation after an interrupt. It creates a
   * tool response part that can be passed to {@link ResumeOptions#getRespond()}.
   *
   * <p>
   * Example usage:
   * 
   * <pre>{@code
   * // Get interrupt from response
   * Part interrupt = response.getInterrupts().get(0);
   * 
   * // Create response with user-provided data
   * Part responseData = tool.respond(interrupt, userConfirmation);
   * 
   * // Resume generation
   * ModelResponse resumed = genkit.generate(GenerateOptions.builder().messages(response.getMessages())
   * 		.resume(ResumeOptions.builder().respond(responseData).build()).build());
   * }</pre>
   *
   * @param interrupt
   *            the interrupted tool request part
   * @param output
   *            the output data to respond with
   * @return a tool response part
   */
  public Part respond(Part interrupt, O output) {
    return respond(interrupt, output, null);
  }

  /**
   * Constructs a tool response for an interrupted tool request with metadata.
   *
   * @param interrupt
   *            the interrupted tool request part
   * @param output
   *            the output data to respond with
   * @param metadata
   *            optional metadata to include in the response
   * @return a tool response part
   */
  public Part respond(Part interrupt, O output, Map<String, Object> metadata) {
    if (interrupt == null || interrupt.getToolRequest() == null) {
      throw new IllegalArgumentException("Interrupt must be a tool request part");
    }

    ToolRequest toolRequest = interrupt.getToolRequest();
    Part responsePart = new Part();
    ToolResponse toolResponse = new ToolResponse(toolRequest.getRef(), toolRequest.getName(), output);
    responsePart.setToolResponse(toolResponse);

    // Add interruptResponse marker in metadata
    Map<String, Object> responseMetadata = new HashMap<>();
    responseMetadata.put("interruptResponse", true);
    if (metadata != null) {
      responseMetadata.putAll(metadata);
    }
    responsePart.setMetadata(responseMetadata);

    return responsePart;
  }

  /**
   * Constructs a restart request for an interrupted tool.
   *
   * <p>
   * This method creates a tool request that will cause the tool to be
   * re-executed. The resumed metadata will be passed to the tool handler.
   *
   * <p>
   * Example usage:
   * 
   * <pre>{@code
   * // Get interrupt from response
   * Part interrupt = response.getInterrupts().get(0);
   * 
   * // Create restart request with confirmation metadata
   * Part restartRequest = tool.restart(interrupt, Map.of("confirmed", true));
   * 
   * // Resume generation
   * ModelResponse resumed = genkit.generate(GenerateOptions.builder().messages(response.getMessages())
   * 		.resume(ResumeOptions.builder().restart(restartRequest).build()).build());
   * }</pre>
   *
   * @param interrupt
   *            the interrupted tool request part
   * @param resumedMetadata
   *            metadata to pass to the tool handler's resumed context
   * @return a tool request part for restart
   */
  public Part restart(Part interrupt, Map<String, Object> resumedMetadata) {
    return restart(interrupt, resumedMetadata, null);
  }

  /**
   * Constructs a restart request with replacement input.
   *
   * @param interrupt
   *            the interrupted tool request part
   * @param resumedMetadata
   *            metadata to pass to the tool handler's resumed context
   * @param replaceInput
   *            optional new input to use instead of the original
   * @return a tool request part for restart
   */
  public Part restart(Part interrupt, Map<String, Object> resumedMetadata, I replaceInput) {
    if (interrupt == null || interrupt.getToolRequest() == null) {
      throw new IllegalArgumentException("Interrupt must be a tool request part");
    }

    ToolRequest originalRequest = interrupt.getToolRequest();
    Part restartPart = new Part();

    // Create new tool request with either original or replacement input
    Object inputToUse = replaceInput != null ? replaceInput : originalRequest.getInput();
    ToolRequest restartRequest = new ToolRequest(originalRequest.getName(), inputToUse);
    restartRequest.setRef(originalRequest.getRef());
    restartPart.setToolRequest(restartRequest);

    // Add resumed metadata
    Map<String, Object> restartMetadata = new HashMap<>();
    restartMetadata.put("source", "restart");
    if (resumedMetadata != null) {
      restartMetadata.put("resumed", resumedMetadata);
    } else {
      restartMetadata.put("resumed", true);
    }
    restartPart.setMetadata(restartMetadata);

    return restartPart;
  }

  /**
   * Builder for Tool.
   *
   * @param <I>
   *            the input type
   * @param <O>
   *            the output type
   */
  public static class Builder<I, O> {
    private String name;
    private String description;
    private Map<String, Object> inputSchema;
    private Map<String, Object> outputSchema;
    private Class<I> inputClass;
    private BiFunction<ActionContext, I, O> handler;

    public Builder<I, O> name(String name) {
      this.name = name;
      return this;
    }

    public Builder<I, O> description(String description) {
      this.description = description;
      return this;
    }

    public Builder<I, O> inputSchema(Map<String, Object> inputSchema) {
      this.inputSchema = inputSchema;
      return this;
    }

    public Builder<I, O> outputSchema(Map<String, Object> outputSchema) {
      this.outputSchema = outputSchema;
      return this;
    }

    public Builder<I, O> inputClass(Class<I> inputClass) {
      this.inputClass = inputClass;
      return this;
    }

    public Builder<I, O> handler(BiFunction<ActionContext, I, O> handler) {
      this.handler = handler;
      return this;
    }

    public Tool<I, O> build() {
      if (name == null) {
        throw new IllegalStateException("Tool name is required");
      }
      if (handler == null) {
        throw new IllegalStateException("Tool handler is required");
      }
      return new Tool<>(name, description, inputSchema, outputSchema, inputClass, handler);
    }
  }
}
